File: src/zros_cpp/impl/subscriber_impl.cc
//
// Created by pairs on 7/28/18.
//

#include "subscriber_impl.h"
#include "zeromq_help.h"

namespace zros {

    ZmqSubStub::ZmqSubStub(const std::string &topic, MessageCb cb):
            ISubStub(topic),
            message_cb_(cb),
            running_(false),
            stop_(false) {
        context_ = std::make_shared<zmq::context_t>(1);
        subscriber_ = std::make_shared<zmq::socket_t>(*context_.get(), ZMQ_SUB);
    }

    bool ZmqSubStub::start() {
        if (running_) {
            SSPD_LOG_INFO << "the topic: " << topic_ << "was already recv message";
            return true;
        }
        subscriber_->setsockopt(ZMQ_SUBSCRIBE, NULL, 0);
        recvThread_ = std::make_shared<std::thread>(&ZmqSubStub::recvMessage, this);
        running_ = true;
        return true;
    }

    bool ZmqSubStub::stop() {
        stop_ = true;
        running_ = false;
        return true;
    }

    bool ZmqSubStub::addPublisher(const std::string &address) {
        std::lock_guard<std::mutex> lk(pub_mutex_);
        if (publishers_address_.find(address) == publishers_address_.end()) {
            subscriber_->connect(address.c_str());
            publishers_address_[address] = 1;
        }
        start();
        return true;
    }

    bool ZmqSubStub::removePublisher(const std::string &address) {
        std::lock_guard<std::mutex> lk(pub_mutex_);
        if (publishers_address_.find(address) != publishers_address_.end()) {
            subscriber_->unbind(address);
            publishers_address_.erase(address);
        }
        return true;
    }

    void ZmqSubStub::recvMessage() {
        while(!stop_) {
            try {
                std::string data;
                // blocking waiting publish data
                data = zeromq::ZmqHelp::s_recv(*(subscriber_.get()));
                message_cb_(data);
            } catch(std::exception& e) {
                SSPD_LOG_ERROR << "zmq recv message thread throw exception, because of " << e.what();
            }
        }
    }

    ZmqSubStub::~ZmqSubStub() {
        stop();
        if (recvThread_) {
            recvThread_->join();
        }
    }

    ////////////// subscriberImpl ///////////////////
    SubscribersImpl::SubscribersImpl() {

    }

    void SubscribersImpl::registerPublisher(const zros_rpc::PublisherInfo *publisherInfo, zros_rpc::Status *status) {
        std::lock_guard<std::mutex> lk(subscriber_mutex_);
        auto topic = publisherInfo->topic();
        if (subscribers_.find(topic) == subscribers_.end()) {
            SSPD_LOG_WARNING << "please check registerPublisher logical error " << topic;
            status->set_code(status->NOT_FOUND);
            status->set_details(topic + " publisher not register on");
            return;
        }
        auto publish_address = publisherInfo->physical_node_info().real_address();
        auto pair = subscribers_[topic];
        auto stub = pair.second;
        stub->addPublisher(publish_address);
        status->set_code(status->OK);
        SSPD_LOG_INFO << "register publisher to : " << topic << " on address: " << publish_address << " success";
    }

    void SubscribersImpl::unregisterPublisher(const zros_rpc::PublisherInfo *publisherInfo, zros_rpc::Status *status) {

    }

    bool SubscribersImpl::registerSubscriber(std::shared_ptr<ISubscriber> subscriber) {
        std::lock_guard<std::mutex> lk(subscriber_mutex_);
        auto topic = subscriber->get_topic();
        if (subscribers_.find(topic) != subscribers_.end()) {
            SSPD_LOG_WARNING << "please check this subscriber already register " << topic;
            return false;
        }
        auto stub = std::make_shared<ZmqSubStub>(topic, [subscriber](const std::string& message){
            subscriber->handleRawMessage(message);
        });
        auto pair = std::make_pair<std::shared_ptr<ISubscriber>, std::shared_ptr<ZmqSubStub> >(std::move(subscriber), std::move(stub));
        subscribers_[topic] = pair;
        return true;
    }

    bool SubscribersImpl::unregisterSubscriber(const std::string &topic) {
        std::lock_guard<std::mutex> lk(subscriber_mutex_);
        subscribers_.erase(topic);
        return true;
    }
}   // namespace zros

File: src/zros_python/impl/subscriber_impl.py
# -*- coding: utf-8 -*-
""" impl subscriber management

data: 2018-08-02
author: pairs
"""
import abc
import logging
import zmq
import threading
import zros_python.pb.zros_pb2 as zpb2

logger = logging.getLogger(__name__)


class ISubStub(object):
    __metaclass__ = abc.ABCMeta

    def __init__(self, topic):
        self._topic = topic
        self._pub_mutex = threading.Lock()
        self._publishers_address = {}

    @abc.abstractmethod
    def start(self):
        pass

    @abc.abstractmethod
    def stop(self):
        pass

    @abc.abstractmethod
    def add_publisher(self, address):
        pass

    @abc.abstractmethod
    def remove_publisher(self, address):
        pass


class ZmqStubStub(ISubStub):
    """

    """
    def __init__(self, topic, callback):
        super(ZmqStubStub, self).__init__(topic)
        self._callback = callback
        self._context = zmq.Context()
        self._subscriber = self._context.socket(zmq.SUB)
        self._stop_mutex = threading.Lock()
        self._stop = False

        self._running_mutex = threading.Lock()
        self._running = False
        self._receive_thread = threading.Thread(target=ZmqStubStub._receive_message, args=(self,))

    def start(self):
        with self._running_mutex:
            if self._running:
                logger.warning(u'the topic %s was already receive message')
                return True
            self._subscriber.setsockopt(zmq.SUBSCRIBE, '')
            self._receive_thread.daemon = True
            self._receive_thread.start()
            self._running = True

    def stop(self):
        with self._running_mutex:
            with self._stop_mutex:
                self._stop = True
                self._running = False
        return True

    def add_publisher(self, address):
        with self._pub_mutex:
            if address not in self._publishers_address:
                self._subscriber.connect(address)
                self._publishers_address[address] = 1
        self.start()
        return True

    def remove_publisher(self, address):
        with self._pub_mutex:
            if address in self._publishers_address:
                del self._publishers_address[address]
        pass

    def _receive_message(self):
        while True:
            with self._stop_mutex:
                if self._stop is True:
                    break
                else:
                    data = self._subscriber.recv()
                    self._callback(data)


class SubscribersImpl(object):
    """

    """
    def __init__(self):
        self._subscriber_mutex = threading.Lock()
        self._subscribers = {}  # key: topic, value:(subscriber, stub)

    def register_publisher(self, publisher_info):
        topic = publisher_info.topic
        status = zpb2.Status()
        with self._subscriber_mutex:
            if topic not in self._subscribers:
                logger.warning(u'please check register publisher %s logical error ', topic)
                status.code = zpb2.Status.NOT_FOUND
                status.details = topic + u' publisher not register on'
                return status

            publish_address = publisher_info.physical_node_info.real_address
            pair = self._subscribers[topic]
            stub = pair[1]
            stub.add_publisher(publish_address)
            status.code = zpb2.Status.OK
            logger.info(u'register publisher to : %s topic on address %s success', topic, publish_address)
            return status

    def unregister_publisher(self, publisher_info):
        pass

    def register_subscriber(self, subscriber):
        topic = subscriber.get_topic()
        with self._subscriber_mutex:
            if topic in self._subscribers:
                logger.warning(u'please check this subscriber already register %s ', topic)
                return False
            stub = ZmqStubStub(topic, subscriber.handle_raw_message)
            pair = (subscriber, stub)
            self._subscribers[topic] = pair
            return True

    def unregister_subscriber(self, topic):
        with self._subscriber_mutex:
            if topic in self._subscribers:
                del self._subscribers[topic]
